### ISSUE TITLE
Implement Robust DPatch Attack

### ISSUE THREAD

**Issue by deprit** [twosixlabs/armory-library#149]
Armory Library must support custom evasion attacks designed by the user; supporting user-developed attacks requires the steps below.
- Design and implement a custom attack API that
    - Optimizes an attack over a training split and, possibly, a validation set;   
    - Exposes the optimization loop to PyTorch Lightning;
    - Logs optimization metrics and attack artifiacts to MLFlow.
- Demonstrate a custom attack that 
    - Implements [Robust DPatch](https://arxiv.org/abs/1906.11897) for object detection ([ART implementation](https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/art/attacks/evasion/dpatch_robust.py));
    - Optimizes the Robust DPatch attack on the [VisDrone dataset ](https://github.com/VisDrone/VisDrone-Dataset)
       using a pre-trained [Hugging Face object detector](https://huggingface.co/smidm/yolov5-visdrone).

**Comment by treubig26** [comment]
This has been partially addressed by PR https://github.com/twosixlabs/armory-library/pull/155, however more work remains to generalize and integrate the attack optimization as part of the armory-library API.

Specifically, a new attack optimization engine should be created that will handle instantiation of the lightning module (with more hooks/parameterization) and the trainer.

The following comments from the PR should also be addressed in a follow-on PR:
- https://github.com/twosixlabs/armory-library/pull/155#discussion_r1626296941
- https://github.com/twosixlabs/armory-library/pull/155#discussion_r1626301511
- https://github.com/twosixlabs/armory-library/pull/155#discussion_r1626364323
- https://github.com/twosixlabs/armory-library/pull/155#discussion_r1626371175
- https://github.com/twosixlabs/armory-library/pull/155#discussion_r1626376518
- https://github.com/twosixlabs/armory-library/pull/155#discussion_r1628036045
